File: drivers/java/src/main/java/com/rethinkdb/net/Handshake.java
package com.rethinkdb.net;

import com.rethinkdb.gen.exc.ReqlAuthError;
import com.rethinkdb.gen.exc.ReqlDriverError;
import com.rethinkdb.gen.proto.Protocol;
import com.rethinkdb.gen.proto.Version;
import org.json.simple.JSONObject;

import java.nio.ByteBuffer;
import java.security.MessageDigest;
import java.util.Optional;

import static com.rethinkdb.net.Crypto.*;
import static com.rethinkdb.net.Util.toJSON;
import static com.rethinkdb.net.Util.toUTF8;

public class Handshake {
    static final Version VERSION = Version.V1_0;
    static final Long SUB_PROTOCOL_VERSION = 0L;
    static final Protocol PROTOCOL = Protocol.JSON;


    private static final String CLIENT_KEY = "Client Key";
    private static final String SERVER_KEY = "Server Key";

    private final String username;
    private final String password;
    private ProtocolState state;

    private interface ProtocolState {
        ProtocolState nextState(String response);
        Optional<ByteBuffer> toSend();
        boolean isFinished();
    }

    private class InitialState implements ProtocolState {
        private final String nonce;
        private final String username;
        private final byte[] password;

        InitialState(String username, String password) {
            this.username = username;
            this.password = toUTF8(password);
            this.nonce = makeNonce();
        }

        @Override
        public ProtocolState nextState(String response) {
            if (response != null) {
                throw new ReqlDriverError("Unexpected response");
            }
            // We could use a json serializer, but it's fairly straightforward
            ScramAttributes clientFirstMessageBare = ScramAttributes.create()
                    .username(username)
                    .nonce(nonce);
            byte[] jsonBytes = toUTF8(
                "{" +
                    "\"protocol_version\":" + SUB_PROTOCOL_VERSION + "," +
                    "\"authentication_method\":\"SCRAM-SHA-256\"," +
                    "\"authentication\":" + "\"n,," + clientFirstMessageBare + "\"" +
                "}"
            );
            ByteBuffer msg = Util.leByteBuffer(
                    Integer.BYTES +    // size of VERSION
                    jsonBytes.length + // json auth payload
                    1                  // terminating null byte
            ).putInt(VERSION.value)
             .put(jsonBytes)
             .put(new byte[1]);
            return new WaitingForProtocolRange(
                    nonce, password, clientFirstMessageBare, msg);
        }

        @Override
        public Optional<ByteBuffer> toSend() {
            return Optional.empty();
        }

        @Override
        public boolean isFinished() {
            return false;
        }
    }

    private class WaitingForProtocolRange implements ProtocolState {
        private final String nonce;
        private final ByteBuffer message;
        private final ScramAttributes clientFirstMessageBare;
        private final byte[] password;

        WaitingForProtocolRange(
                String nonce,
                byte[] password,
                ScramAttributes clientFirstMessageBare,
                ByteBuffer message) {
            this.nonce = nonce;
            this.password = password;
            this.clientFirstMessageBare = clientFirstMessageBare;
            this.message = message;
        }

        @Override
        public ProtocolState nextState(String response) {
            JSONObject json = toJSON(response);
            throwIfFailure(json);
            Long minVersion = (Long) json.get("min_protocol_version");
            Long maxVersion = (Long) json.get("max_protocol_version");
            if (SUB_PROTOCOL_VERSION < minVersion || SUB_PROTOCOL_VERSION > maxVersion) {
                throw new ReqlDriverError(
                        "Unsupported protocol version " + SUB_PROTOCOL_VERSION +
                                ", expected between " + minVersion + " and " + maxVersion);
            }
            return new WaitingForAuthResponse(nonce, password, clientFirstMessageBare);
        }

        @Override
        public Optional<ByteBuffer> toSend() {
            return Optional.of(message);
        }

        @Override
        public boolean isFinished() {
            return false;
        }
    }

    private class WaitingForAuthResponse implements ProtocolState {
        private final String nonce;
        private final byte[] password;
        private final ScramAttributes clientFirstMessageBare;

        WaitingForAuthResponse(
                String nonce, byte[] password, ScramAttributes clientFirstMessageBare) {
            this.nonce = nonce;
            this.password = password;
            this.clientFirstMessageBare = clientFirstMessageBare;
        }

        @Override
        public ProtocolState nextState(String response) {
            JSONObject json = toJSON(response);
            throwIfFailure(json);
            String serverFirstMessage = (String) json.get("authentication");
            ScramAttributes serverAuth = ScramAttributes.from(serverFirstMessage);
            if (!serverAuth.nonce().startsWith(nonce)) {
                throw new ReqlAuthError("Invalid nonce from server");
            }
            ScramAttributes clientFinalMessageWithoutProof = ScramAttributes.create()
                    .headerAndChannelBinding("biws")
                    .nonce(serverAuth.nonce());

            // SaltedPassword := Hi(Normalize(password), salt, i)
            byte[] saltedPassword = pbkdf2(
                    password, serverAuth.salt(), serverAuth.iterationCount());

            // ClientKey := HMAC(SaltedPassword, "Client Key")
            byte[] clientKey = hmac(saltedPassword, CLIENT_KEY);

            // StoredKey := H(ClientKey)
            byte[] storedKey = sha256(clientKey);

            // AuthMessage := client-first-message-bare + "," +
            //                server-first-message + "," +
            //                client-final-message-without-proof
            String authMessage =
                    clientFirstMessageBare + "," +
                    serverFirstMessage + "," +
                    clientFinalMessageWithoutProof;

            // ClientSignature := HMAC(StoredKey, AuthMessage)
            byte[] clientSignature = hmac(storedKey, authMessage);

            // ClientProof := ClientKey XOR ClientSignature
            byte[] clientProof = xor(clientKey, clientSignature);

            // ServerKey := HMAC(SaltedPassword, "Server Key")
            byte[] serverKey = hmac(saltedPassword, SERVER_KEY);

            // ServerSignature := HMAC(ServerKey, AuthMessage)
            byte[] serverSignature = hmac(serverKey, authMessage);

            ScramAttributes auth = clientFinalMessageWithoutProof
                    .clientProof(clientProof);
            byte[] authJson = toUTF8("{\"authentication\":\"" + auth + "\"}");
            ByteBuffer message = Util.leByteBuffer(authJson.length + 1)
                    .put(authJson)
                    .put(new byte[1]);
            return new WaitingForAuthSuccess(serverSignature, message);
        }

        @Override
        public Optional<ByteBuffer> toSend() {
            return Optional.empty();
        }

        @Override
        public boolean isFinished() {
            return false;
        }
    }

    private class WaitingForAuthSuccess implements ProtocolState {
        private final byte[] serverSignature;
        private final ByteBuffer message;

        public WaitingForAuthSuccess(byte[] serverSignature, ByteBuffer message) {
            this.serverSignature = serverSignature;
            this.message = message;
        }

        @Override
        public ProtocolState nextState(String response) {
            JSONObject json = toJSON(response);
            throwIfFailure(json);
            ScramAttributes auth = ScramAttributes
                    .from((String) json.get("authentication"));
            if (!MessageDigest.isEqual(auth.serverSignature(), serverSignature)) {
                throw new ReqlAuthError("Invalid server signature");
            }
            return new HandshakeSuccess();
        }

        @Override
        public Optional<ByteBuffer> toSend() {
            return Optional.of(message);
        }

        @Override
        public boolean isFinished() {
            return false;
        }
    }

    private class HandshakeSuccess implements ProtocolState {

        @Override
        public ProtocolState nextState(String response) {
            return this;
        }

        @Override
        public Optional<ByteBuffer> toSend() {
            return Optional.empty();
        }

        @Override
        public boolean isFinished() {
            return true;
        }
    }

    private void throwIfFailure(JSONObject json) {
        if (!(boolean) json.get("success")) {
            Long errorCode = (Long) json.get("error_code");
            if (errorCode >= 10 && errorCode <= 20) {
                throw new ReqlAuthError((String) json.get("error"));
            } else {
                throw new ReqlDriverError((String) json.get("error"));
            }
        }
    }

    public Handshake(String username, String password) {
        this.username = username;
        this.password = password;
        this.state = new InitialState(username, password);
    }

    public void reset() {
        this.state = new InitialState(this.username, this.password);
    }

    public Optional<ByteBuffer> nextMessage(String response) {
        this.state = this.state.nextState(response);
        return this.state.toSend();
    }

    public boolean isFinished() {
        return this.state.isFinished();
    }
}

File: drivers/python/rethinkdb/handshake.py
import base64
import binascii
import hashlib
import hmac
import random
import struct
import sys
import threading

from . import ql2_pb2
from .errors import *

try:
    xrange
except NameError:
    xrange = range

class HandshakeV0_4(object):
    VERSION = ql2_pb2.VersionDummy.Version.V0_4
    PROTOCOL = ql2_pb2.VersionDummy.Protocol.JSON

    def __init__(self, host, port, auth_key):
        self._host = host
        self._port = port
        self._auth_key = auth_key.encode("ascii")

        self._state = 0

    def reset(self):
        self._state = 0

    def next_message(self, response):
        if self._state == 0:
            if response is not None:
                raise ReqlDriverError("Unexpected response")
            self._state = 1
            return \
                struct.pack("<2L", self.VERSION, len(self._auth_key)) + \
                self._auth_key + \
                struct.pack("<L", self.PROTOCOL)
        elif self._state == 1:
            if response is None:
                raise ReqlDriverError("Expected response")
            self._state = 2
            if response != b"SUCCESS":
                # This is an error case, we attempt to decode `response` as UTF-8 with
                # with fallbacks to get something useful
                message = None
                try:
                    message = response.decode("utf-8", errors="ignore").strip()
                except TypeError:
                    try:
                        message = response.decode("utf-8").strip()
                    except UnicodeError:
                        message = repr(response).strip()
                if message == "ERROR: Incorrect authorization key.":
                    raise ReqlAuthError("Incorrect authentication key.", self._host, self._port)
                else:
                    raise ReqlDriverError(
                        "Server dropped connection with message: \"%s\"" % (message, ))
            return None
        else:
            raise ReqlDriverError("Unexpected handshake state")

class HandshakeV1_0(object):
    VERSION = ql2_pb2.VersionDummy.Version.V1_0
    PROTOCOL = ql2_pb2.VersionDummy.Protocol.JSON

    def __init__(self, json_decoder, json_encoder, host, port, username, password):
        self._json_decoder = json_decoder
        self._json_encoder = json_encoder
        self._host = host
        self._port = port
        self._username = \
            username.encode("utf-8").replace(b"=", b"=3D").replace(b",", b"=2C")
        try:
            self._password = bytes(password, "utf-8")
        except TypeError:
            self._password = bytes(password)

        # Python 2.7.7+ and Python 3.3+
        self._compare_digest = getattr(hmac, "compare_digest", None)
        if not self._compare_digest:
            self._compare_digest = self.__compare_digest

        # Python 2.7.8+ and Python 3.4+.
        self._pbkdf2_hmac = getattr(hashlib, "pbkdf2_hmac", None)
        if not self._pbkdf2_hmac:
            self._pbkdf2_hmac = HandshakeV1_0.__pbkdf2_hmac

        self._protocol_version = 0
        self._random = random.SystemRandom()
        self._state = 0

    def reset(self):
        self._r = None
        self._client_first_message_bare = None
        self._server_signature = None
        self._state = 0

    def next_message(self, response):
        if self._state == 0:
            if response is not None:
                raise ReqlDriverError("Unexpected response")

            # Using base64 encoding for printable characters
            self._r = base64.standard_b64encode(
                bytes(bytearray(self._random.getrandbits(8) for i in xrange(18))))

            self._client_first_message_bare = b"n=" + self._username + b",r=" + self._r

            # Here we send the version as well as the initial JSON as an optimization
            self._state = 1
            return struct.pack("<L", self.VERSION) + \
                self._json_encoder.encode({
                    "protocol_version": self._protocol_version,
                    "authentication_method": "SCRAM-SHA-256",
                    "authentication":
                        (b"n,," + self._client_first_message_bare).decode("ascii")
                }).encode("utf-8") + \
                b'\0'
        elif self._state == 1:
            response = response.decode("utf-8")
            if response.startswith("ERROR"):
                raise ReqlDriverError("Received an unexpected reply.  You may be attempting to connect to a RethinkDB server that is too old for this driver.  The minimum supported server version is 2.3.0.")
            json = self._json_decoder.decode(response)
            try:
                if json["success"] == False:
                    if 10 <= json["error_code"] <= 20:
                        raise ReqlAuthError(json["error"], self._host, self._port)
                    else:
                        raise ReqlDriverError(json["error"])

                min = json["min_protocol_version"]
                max = json["max_protocol_version"]
                if not min <= self._protocol_version <= max:
                    raise ReqlDriverError(
                        "Unsupported protocol version %d, expected between %d and %d" % (
                            self._protocol_version, min, max))
            except KeyError as key_error:
                raise ReqlDriverError("Missing key: %s" % (key_error, ))

            # We've already sent the initial JSON above, and only support a single
            # protocol version at the moment thus we simply read the next response
            self._state = 2
            return ""
        elif self._state == 2:
            json = self._json_decoder.decode(response.decode("utf-8"))
            server_first_message = r = salt = i = None
            try:
                if json["success"] == False:
                    if 10 <= json["error_code"] <= 20:
                        raise ReqlAuthError(json["error"], self._host, self._port)
                    else:
                        raise ReqlDriverError(json["error"])

                server_first_message = json["authentication"].encode("ascii")
                authentication = dict(
                    x.split(b"=", 1) for x in server_first_message.split(b","))

                r = authentication[b"r"]
                if not r.startswith(self._r):
                    raise ReqlAuthError("Invalid nonce from server", self._host, self._port)
                salt = base64.standard_b64decode(authentication[b"s"])
                i = int(authentication[b"i"])
            except KeyError as key_error:
                raise ReqlDriverError("Missing key: %s" % (key_error, ))

            client_final_message_without_proof = b"c=biws,r=" + r

            # SaltedPassword := Hi(Normalize(password), salt, i)
            salted_password = self._pbkdf2_hmac("sha256", self._password, salt, i)

            # ClientKey := HMAC(SaltedPassword, "Client Key")
            client_key = hmac.new(
                salted_password, b"Client Key", hashlib.sha256).digest()

            # StoredKey := H(ClientKey)
            stored_key = hashlib.sha256(client_key).digest()

            # AuthMessage := client-first-message-bare + "," +
            #                server-first-message + "," +
            #                client-final-message-without-proof
            auth_message = b",".join((
                self._client_first_message_bare,
                server_first_message,
                client_final_message_without_proof))

            # ClientSignature := HMAC(StoredKey, AuthMessage)
            client_signature = hmac.new(
                stored_key, auth_message, hashlib.sha256).digest()

            # ClientProof := ClientKey XOR ClientSignature
            client_proof = struct.pack(
                "32B",
                *(l ^ r for l, r in zip(
                    struct.unpack("32B", client_key),
                    struct.unpack("32B", client_signature))))

            # ServerKey := HMAC(SaltedPassword, "Server Key")
            server_key = hmac.new(
                salted_password, b"Server Key", hashlib.sha256).digest()

            # ServerSignature := HMAC(ServerKey, AuthMessage)
            self._server_signature = hmac.new(
                server_key, auth_message, hashlib.sha256).digest()

            self._state = 3
            return self._json_encoder.encode({
                "authentication": (
                        client_final_message_without_proof +
                        b",p=" + base64.standard_b64encode(client_proof)
                    ).decode("ascii")
            }).encode("utf-8") + \
            b'\0'
        elif self._state == 3:
            json = self._json_decoder.decode(response.decode("utf-8"))
            v = None
            try:
                if json["success"] == False:
                    if 10 <= json["error_code"] <= 20:
                        raise ReqlAuthError(json["error"], self._host, self._port)
                    else:
                        raise ReqlDriverError(json["error"])

                authentication = dict(
                    x.split(b"=", 1)
                        for x in json["authentication"].encode("ascii").split(b","))

                v = base64.standard_b64decode(authentication[b"v"])
            except KeyError as key_error:
                raise ReqlDriverError("Missing key: %s" % (key_error, ))

            if not self._compare_digest(v, self._server_signature):
                raise ReqlAuthError("Invalid server signature", self._host, self._port)

            self._state = 4
            return None
        else:
            raise ReqlDriverError("Unexpected handshake state");

    @staticmethod
    def __compare_digest(a, b):
        if sys.version_info[0] == 3:
            def xor_bytes(a, b):
                return a ^ b
        else:
            def xor_bytes(a, b, _ord=ord):
                return _ord(a) ^ _ord(b)

        left = None
        right = b
        if len(a) == len(b):
            left = a
            result = 0
        if len(a) != len(b):
            left = b
            result = 1

        for l, r in zip(left, right):
            result |= xor_bytes(l, r)

        return result == 0

    class thread_local_cache(threading.local):
        def __init__(self):
            self.cache = {}

        def set(self, key, val):
            self.cache[key] = val

        def get(self, key):
            return self.cache.get(key)

    pbkdf2_cache = thread_local_cache()

    @staticmethod
    def __pbkdf2_hmac(hash_name, password, salt, iterations):
        assert hash_name == "sha256", hash_name

        def from_bytes(value, hexlify=binascii.hexlify, int=int):
            return int(hexlify(value), 16)

        def to_bytes(value, unhexlify=binascii.unhexlify):
            try:
                return unhexlify(bytes("%064x" % value, "ascii"))
            except TypeError:
                return unhexlify(bytes("%064x" % value))

        cache_key = (password, salt, iterations)

        cache_result = HandshakeV1_0.pbkdf2_cache.get(cache_key)

        if cache_result is not None:
            return cache_result

        mac = hmac.new(password, None, hashlib.sha256)

        def digest(msg, mac=mac):
            mac_copy = mac.copy()
            mac_copy.update(msg)
            return mac_copy.digest()

        t = digest(salt + b"\x00\x00\x00\x01")
        u = from_bytes(t)
        for c in xrange(iterations - 1):
            t = digest(t)
            u ^= from_bytes(t)

        u = to_bytes(u)
        HandshakeV1_0.pbkdf2_cache.set(cache_key, u)
        return u
